static/js/compose_ui.js

```diff
@@ -27,16 +27,19 @@ exports.smart_insert = function (textarea, syntax) {
         }
     }
 
-    textarea.caret(syntax);
     textarea.focus();
+    return syntax;
 };
 
 exports.insert_syntax_and_focus = function (syntax) {
     // Generic helper for inserting syntax into the main compose box
     // where the cursor was and focusing the area.  Mostly a thin
     // wrapper around smart_insert.
     var textarea = $('#new_message_content');
-    exports.smart_insert(textarea, syntax);
+    syntax = exports.smart_insert(textarea, syntax);
+    if (!document.execCommand("insertText", false, syntax)) {
+        textarea.caret(syntax);
+    }
```

Comment:
@YJDave I realize I suggested putting the changes in here, but obviously it makes more sense to put this in `smart_insert`, since these changes have nothing to do with focusing the textbox.  I'd encourage you to think through these things rather than taking my word for it.

Meanwhile, @showell @brockwhittaker can you review whether this `insertText` thing is the best way to implement the "undo" functionality?  It apparently works well in Chrome but not Firefox.

Reply:
oh, yeah!! sorry for that. Will update soon.